### PR TITLE
Use `napari/packaging` for Conda CI

### LINKED ISSUE #5: Migrate CI infrastructure from `napari/napari`
- [x] Nightly conda packages
- [x] Installers
- [x] Signing, notarization and stapling
- [x] Installer validation
- [x] In-repo artifact publishing
- [x] Cross-repo triggers for releases (final and RCs)
- [x] Cross-repo artifact publishing

### PR DESCRIPTION
**Pull request by jaimergp**
# Description

Following NAP-2, we are migrating the CI for packaging over to `napari/packaging`. However, due to the required cross-repository integrations, we will still run it from `napari/napari` thanks to the `workflow_call` triggers.

This allows to reuse the remote workflow with a `napari/napari` token, essentially. That way artifact publishing is still kept in this repo.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# References

Closes https://github.com/napari/packaging/issues/5

# How has this been tested?

- [x] CI functionality remains the same as it was before.

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
If we need to be able to build again some old version of napari after some delay (longer than log retention)? If yes, then we need some option to determine `napari/packaging` version that should be used. 

Or it should be enough universal?

**Comment by jaimergp** [comment]
> If we need to be able to build again some old version of napari after some delay (longer than log retention)?

I don't think we are aiming for perfect reproducibility. The conda env solve will be different to begin with. _If_ that's ever needed, we can roll back in commit history so the dates match and use the commit hash in the `uses` keyword (after the `@`).

**Comment by liu-ziyang** [comment]
Outside scope of this PR, but I wish we could drop conda building from push commits and PRs. The daily runs are good enough to verify we are not breaking things. Right now the false negative pretty much defeated the point of running on individual commit or PR

**Comment by jaimergp** [comment]
That's totally ok. The YML file on this PR now basically list the triggers, so the team can focus on which ones are needed :)

**Comment by jaimergp** [comment]
Commit [bd6bffa](https://github.com/napari/napari/pull/4923/commits/bd6bffa23111be1bec45a11185a18196871bf5bc) simulates full-signing/notarization/stapling capabilities once merged. Everything works, including uploads to Anaconda.org and CI artifacts.

* Artifacts: https://github.com/napari/napari/actions/runs/2845608450
* Anaconda.org uploads: https://anaconda.org/napari/napari/files?version=0.4.16rc2.dev228%2Bg8c302ed7


I'll now revert to reduced triggers, only for scheduled runs and tagged versions.

**Comment by jaimergp** [comment]
Ready for review, team!

File: .github/workflows/make_bundle_conda.yml [hunk 1]
```diff
@@ -1,438 +1,17 @@
...
-      # - name: Update latest tag
-      #   uses: EndBug/latest-tag@latest
-      #   if: ${{ github.event_name == 'schedule' }}
-      #   with:
-      #     description: latest code released from nightly build
-      #     tag-name: latest
-      #   env:
-      #     GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  packaging:
+    uses: napari/packaging/.github/workflows/make_bundle_conda.yml@main
```
Comment (tlambert03) [inline .github/workflows/make_bundle_conda.yml:14]: nifty

**Review by tlambert03 — APPROVED** [review]
looks good to me!

complete side-question: would it be possible to add a couple lines to `packaging.md` detailing how to test the constructor build locally?  Is it as simple as running `python bundle_conda.py`?  Or does one need to carefully construct an environment with variables first?

**Review by liu-ziyang — APPROVED** [review]
LGTM, some minor nit pickings

File: docs/developers/packaging.md [hunk 2]
```diff
@@ -82,35 +93,43 @@ supports the following installer types:
...
-  - conda  # we add these to assist in the plugin installations
-  - mamba  # we add these to assist in the plugin installations
-  - pip    # we add these to assist in the plugin installations
+specs: # specs for the 'base'  environment
+  - python   # pinned to the version of the running interpreter, configured in the CI
+  - conda    # we add these to manage different napari versions
+  - mamba    # we add these to manage different napari versions
+  - pip      # we add these to manage different napari versions
+extra_envs:
+  napari-0.4.16: # this is the environment that will actually contain the napari packages
```
Comment (liu-ziyang) [inline docs/developers/packaging.md:116]: if we are moving internal versions to 0.0.1, should we update here as well?

Comment (jaimergp) [reply docs/developers/packaging.md:116]: The env name does contain the napari version it ships. I haven't yet decided (and will probably post an issue asking for comments) about the path naming conventions.

Right it's:

```
napari-installations-0.0.1/envs/napari-0.4.16
\________________________/      \___________/
          base env                napari env
```

So when the user installs/updates to a new version, we would have:

```
napari-installations-0.0.1/
+--> envs/
     +--> napari-0.4.16/
     +--> napari-0.4.17/

```

**Comment by liu-ziyang** [comment]
before we merge, there seems to be some command issue with the "uploading to anaconda" step in the flow, see: https://github.com/napari/packaging/runs/7830323788?check_suite_focus=true

**Comment by jaimergp** [comment]
> before we merge, there seems to be some command issue with the "uploading to anaconda" step in the flow, see: https://github.com/napari/packaging/runs/7830323788?check_suite_focus=true

That's the error we get when the token is missing. Since `ANACONDA_TOKEN` is only available in this repo (`napari/napari`), the workflow will only succeed when called from this repository. See [this run](https://github.com/napari/napari/runs/7804083853?check_suite_focus=true) for an example.

**Comment by liu-ziyang** [comment]
> > before we merge, there seems to be some command issue with the "uploading to anaconda" step in the flow, see: https://github.com/napari/packaging/runs/7830323788?check_suite_focus=true
> 
> That's the error we get when the token is missing. Since `ANACONDA_TOKEN` is only available in this repo (`napari/napari`), the workflow will only succeed when called from this repository. See [this run](https://github.com/napari/napari/runs/7804083853?check_suite_focus=true) for an example.

I see, shall we disable the runs from the packaging repo explicitly somehow?

**Comment by jaimergp** [comment]
The workflow is now configured to only run under these triggers:

```
  pull_request:  # so we can test stuff

  workflow_call:  # so napari/napari works

  workflow_dispatch:  # to run manually
```

**Review by Carreau — APPROVED** [review]

**Comment by Carreau** [comment]
As this is likely to fix CI in many other Pull requests and only touches CI/releasing, I'm going to merge that, and disable mypy in another PR so that we can get CI to be green again.

**Pull request merged**